R/simulationResultsToDataFrame.R

```diff
@@ -0,0 +1,66 @@
+#' Converts a `SimulationResults` objects to a data.frame
+#'
+#' @inheritParams getOutputValues
+#'
+#' @examples
+#' library(ospsuite)
+#'
+#' simPath <- system.file("extdata", "simple.pkml", package = "ospsuite")
+#' sim <- loadSimulation(simPath)
+#'
+#' # Running an individual simulation
+#' # results is an instance of `SimulationResults`
+#' results <- runSimulations(sim)
+#'
+#' # convert to a dataframe
+#' simulationResultsToDataFrame(results)
+#'
+#' @return
+#'
+#' SimulationResults object as data.frame with columns IndividualId, Time, paths,
+#' simulationValues, unit, dimension, TimeUnit, TimeDimension.
+#'
+#' @export
+
+simulationResultsToDataFrame <- function(simulationResults,
+                                         quantitiesOrPaths = NULL,
+                                         population = NULL,
+                                         individualIds = NULL) {
+  # no need to validating the input because this will be done by
+  # getOutputValues() anyways
+  simList <- getOutputValues(
+    simulationResults = simulationResults,
+    quantitiesOrPaths = quantitiesOrPaths,
+    population = population,
+    individualIds = individualIds
+  )
+
+  # convert data to long format with a new column for paths
+  df_data <- simList$data %>%
+    tidyr::pivot_longer(
+      cols = -c("IndividualId", "Time"),
+      names_to = "paths",
+      values_to = "simulationValues"
+    )
+
+  # extract units and dimensions for paths in a separate dataframe
+  # leave out time units and dimensions since it is not a path
+  df_meta <- purrr::imap_dfr(
+    .x  = simList$metaData,
+    .f  = ~as.data.frame(.x, row.names = NULL),
+    .id = "paths"
+  ) %>%
+    dplyr::filter(paths != "Time")
+
+  # combine these dataframes
+  df <- dplyr::left_join(df_data, df_meta, by = "paths")
+
+  # add time unit and dimnensions now to the combined dataframe
+  df <- dplyr::bind_cols(
+    df,
+    as.data.frame(simList$metaData$Time, col.names = c("TimeUnit", "TimeDimension"))
+  )
+
+  # return the combined dataframe
+  return(df)
```

Comment:
So.. why are we putting all back together in one dataframe with one extra column for units... I thought this was something we DID not want to have ? @IndrajeetPatil @PavelBal ?

Reply:
We need both simulated values and metadata in the long dataframe, which can then be used for plotting.

Reply:
Why are you treating time differently? Cannot it be an entry with Path = "Time"?

Reply:
But `Time` is not a path though, i.e. it is not an observed or simulated value, and therefore not a dependent variable. It is an independent variable, and we can't have it in long format. We need it outside because it links the data in long format to each other.

Code for visualizing this data will, I think, make it clear why we can't have it as another level in the `paths` column:

https://github.com/Open-Systems-Pharmacology/OSPSuite-R/issues/701#issuecomment-994751325

Reply:
I see...ok